content/blog/2023-12-18-torcx-deprecation-for-systemd-sysext.md

```diff
@@ -0,0 +1,82 @@
+
+Prior to our adoption of systemd-sysext, all of these OS customisations and extensions were handled by different, separate mechanisms.
+One of these legacy mechanisms - the one that allowed (within tight limits) to customise the container runtime - was [torcx](https://github.com/flatcar/torcx).
+Even though in theory, torcx can handle arbitrary applications, it was only ever used to ship and to support more than one docker version per OS release.
+It's a very specific tool  - it's unlikely you've ever heard of it, let alone used it.
+
+Torcx-managed applications are built into a tarball, which is either added to the base OS image or can be downloaded at provisioning time from a HTTPS source.
+A torcx "manifest" - a JSON definition of what source (local or remote) to use for which torcx-managed application - is shipped with the operating system, and can be overridden by a user-defined manifest at provisioning time, thus allowing customisations.
+Torcx-managed applications are represented in the OS by means of wrapper scripts - conceptionally not unlike the [Debian Alternatives](https://wiki.debian.org/DebianAlternatives) tool, but with scripts instead of symlinks.
+At early boot, torcx unpacks all torcx-managed applications from the sources / versions that have been defined in the manifests.
+The wrapper scripts can then be called to run the respective binaries.
+
+This approach has, compared to systemd-sysext, multiple drawbacks:
+- Wrapper scripts need to be baked into the base OS to represent the binaries of torcx-managed applications.
+  After the OS image is built, it's not possible to add or modify torcx-managed applications.
+  In other words, users are restricted to customising docker versions when using the official Flatcar releases, and required to build their own images (and maintain their own OS update infrastructure) for other modifications.
+- Torcx applications require special build and packaging since binaries and libraries reside in non-standard paths, increasing maintenance effort required to keep apps up to date (as upstream Gentoo ebuilds cannot easily be used).
+- Torcx-managed application packages were extracted (and possibly downloaded) at early boot, prolonging the boot process.
+  Furthermore, extracted binaries were stored in an ephemeral ramdisk, increasing memory footprint.
+- Updating torcx-managed applications required a change of the manifest (to point to the new version) which would either require re-provisioning or modifying the torcx manifest on nodes, causing config drift.
+
+Also, the Flatcar docs have been recommending [downloading plain docker / containerd binaries](https://www.flatcar.org/docs/latest/container-runtimes/use-a-custom-docker-or-containerd-version/) instead of using torcx for customising docker for quite a while.
```

Comment:
```suggestion
Due to these drawbacks, the Flatcar docs have been recommending [downloading plain docker / containerd binaries](https://www.flatcar.org/docs/latest/container-runtimes/use-a-custom-docker-or-containerd-version/) instead of using torcx for customising docker for quite a while.
```